compiler/qsc_rir/src/passes/ssa_transform.rs

```diff
@@ -0,0 +1,343 @@
+                    let new_var = Variable {
+                        variable_id: next_var_id,
+                        ty: operand.get_type(),
+                    };
+                    let phi_node = Instruction::Phi(args, new_var);
+                    block.0.insert(0, phi_node);
+                    var_map_updates.insert(variable_id, Operand::Variable(new_var));
+                    next_var_id = next_var_id.successor();
+                }
+            }
+
+            // Now that the block has finished processing, apply any updates to the variable map,
+            // and use the updated map to propagate variables through the block.
+            for (var_id, operand) in var_map_updates {
+                let var_map = block_var_map
+                    .get_mut(block_id)
+                    .expect("block should have variable map");
+                var_map.entry(var_id).or_insert(operand);
+                map_variable_use_in_block(block, var_map);
+            }
+        }
+    }
+}
+
+// For now, SSA transform assumes the graph is acyclic, so verify that no block has a predecessor with
+// a block id less than itself, which would indicate a cycle.
+fn ensure_acyclic(preds: &IndexMap<BlockId, Vec<BlockId>>) {
+    for (block_id, block_preds) in preds.iter() {
+        assert!(
+            !block_preds.iter().any(|pred| *pred >= block_id),
+            "block {block_id:?} has a cycle in its predecessors"
+        );
+    }
+}
+
+// Remove store instructions and propagate variables through individual blocks.
+// This produces a per-block map of dynamic variables to their values.
+// Any block with a single predecessor inherits that predecessor's mapped variables, since those
+// are live across the block.
+fn map_store_to_dominated_ssa(
+    program: &mut Program,
+    preds: &IndexMap<BlockId, Vec<BlockId>>,
+) -> IndexMap<BlockId, FxHashMap<VariableId, Operand>> {
+    let mut block_var_map = IndexMap::default();
+    for (block_id, block) in program.blocks.iter_mut() {
+        let mut var_map: FxHashMap<VariableId, Operand> = match preds.get(block_id) {
+            Some(block_preds) if block_preds.len() == 1 => {
+                // Any block with a single predecessor inherits those mapped variables.
+                block_var_map
+                    .get(block_preds[0])
+                    .cloned()
+                    .unwrap_or_default()
+            }
+            _ => FxHashMap::default(),
+        };
+        map_variable_use_in_block(block, &mut var_map);
+        block_var_map.insert(block_id, var_map);
+    }
+    block_var_map
+}
+
+// Propagates stored variables through a block, tracking the latest stored value and replacing
+// usage of the variable with the stored value.
+#[allow(clippy::too_many_lines)]
+fn map_variable_use_in_block(block: &mut Block, var_map: &mut impl VariableMapper) {
+    let instrs = block.0.drain(..).collect::<Vec<_>>();
+
+    for instr in instrs {
+        match instr {
```

Comment:
Actually, now that you point that out, I think this can get even shorter. Since we drain the instructions above, the iterator owns them, which means I might be able to edit them in-place and avoid creating a copy. Let me try that...

Reply:
Sweet, the more efficient pattern worked!